### PR TITLE
chore:non keyed stream support for reduce

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

### CODE REVIEW

**Comment by yhl25** [comment]
I have tested this using a simple non-keyed reduce pipeline 
Screenshot 2022-11-30 at 3 14 55 PM

File: pkg/apis/numaflow/v1alpha1/const.go [hunk 1]
```diff
@@ -137,6 +137,9 @@ const (
 	DefaultStoreSyncDuration  = 2 * time.Second        // Default sync duration for pbq
 	DefaultStoreMaxBufferSize = 100000                 // Default buffer size for pbq in bytes
 	DefaultStorePath          = PathPBQMount + "/wals" // Default store path
+
+	// DefaultGlobalKey Default key for non keyed stream
+	DefaultGlobalKey = "KEY_NOT_SPECIFIED"
```
Comment (ashwinidulams) [inline pkg/apis/numaflow/v1alpha1/const.go:142]: can we say `NON_KEYED_STREAM`

Comment (yhl25) [reply pkg/apis/numaflow/v1alpha1/const.go:142]: yes

File: pkg/reduce/reduce.go [hunk 2]
```diff
@@ -76,6 +79,7 @@ func NewDataForward(ctx context.Context,
 		watermarkPublishers: watermarkPublishers,
 		windowingStrategy:   windowingStrategy,
 		log:                 logging.FromContext(ctx),
+		keyed:               vertex.Spec.UDF.GroupBy.Keyed,
```
Comment (ashwinidulams) [inline pkg/reduce/reduce.go:82]: can we make it boolean and do the conversion in reduce?

File: pkg/reduce/reduce.go [hunk 3]
```diff
@@ -148,8 +152,16 @@ func (d *DataForward) forwardAChunk(ctx context.Context) {
 	// fetch watermark using the first element's watermark, because we assign the watermark to all other
 	// elements in the batch based on the watermark we fetch from 0th offset.
 	processorWM := d.watermarkFetcher.GetWatermark(readMessages[0].ReadOffset)
-	for _, m := range readMessages {
-		m.Watermark = time.Time(processorWM)
+	if d.keyed {
```
Comment (ashwinidulams) [inline pkg/reduce/reduce.go:155]: we could push the if/else inside of for loop.

File: pkg/reduce/reduce.go [hunk 4]
```diff
@@ -148,8 +152,16 @@ func (d *DataForward) forwardAChunk(ctx context.Context) {
...
+		for _, m := range readMessages {
+			m.Watermark = time.Time(processorWM)
+		}
+	} else {
+		for _, m := range readMessages {
+			m.Key = dfv1.DefaultGlobalKey
+			m.Message.Key = dfv1.DefaultGlobalKey
```
Comment (ashwinidulams) [inline pkg/reduce/reduce.go:162]: wondering if we really need to change the key associated with the payload (m.Message)

Comment (ashwinidulams) [reply pkg/reduce/reduce.go:162]: i guess this is fine.

File: pkg/apis/numaflow/v1alpha1/const.go [hunk 5]
```diff
@@ -137,6 +137,9 @@ const (
 	DefaultStoreSyncDuration  = 2 * time.Second        // Default sync duration for pbq
 	DefaultStoreMaxBufferSize = 100000                 // Default buffer size for pbq in bytes
 	DefaultStorePath          = PathPBQMount + "/wals" // Default store path
+
+	// DefaultGlobalKey Default key for non keyed stream
+	DefaultGlobalKey = "NON_KEYED_STREAM"
```
Comment (ashwinidulams) [inline pkg/apis/numaflow/v1alpha1/const.go:142]:
```suggestion
	DefaultKeyForNonKeyedData = "NON_KEYED_STREAM"
```

File: pkg/reduce/reduce.go [hunk 6]
```diff
@@ -148,8 +152,16 @@ func (d *DataForward) forwardAChunk(ctx context.Context) {
...
 	processorWM := d.watermarkFetcher.GetWatermark(readMessages[0].ReadOffset)
-	for _, m := range readMessages {
-		m.Watermark = time.Time(processorWM)
+	if d.keyed {
+		for _, m := range readMessages {
+			m.Watermark = time.Time(processorWM)
+		}
+	} else {
+		for _, m := range readMessages {
+			m.Key = dfv1.DefaultGlobalKey
```
Comment (ashwinidulams) [inline pkg/reduce/reduce.go:161]:
```suggestion
			m.Key = dfv1.DefaultKeyForNonKeyedData
```

File: pkg/reduce/reduce.go (hunk 3 above)
Comment (yhl25) [reply pkg/reduce/reduce.go]: done

File: pkg/reduce/reduce.go (hunk 2 above)
Comment (yhl25) [reply pkg/reduce/reduce.go]: done

**Review by ashwinidulams — APPROVED** [review]

**Review by whynowy — APPROVED** [review]

**Pull request merged**